README
opennap v0.48 (BETA)
the open source (TM) napster server
===================================
August 29, 2004

http://opennap-ng.org
IRC: #opennap on EFNet
Email: khaytsus@opennap-ng.org

opennap is an Open Source(TM) server implementation of the popular
Napster protocol.  It is written in ANSI C and was designed to run on Unix
platforms.  So far this server has been tested on:
	* Linux (i386, alpha, sparc, ppc)
	* BSDI
	* FreeBSD
	* Solaris
	* IRIX
	* OS/2
	* Windows 9x/NT/2k/XP (MS VC++, CYGWIN)
Minimal porting should be required for compilation on other Unix variants
(if you do a port, please send patches to khaytsus@users.sourceforge.net)

Since Napster is not an open specification, much of the internals of the
servers are specific to this implementation.  In particular, I don't expect
that this server will interoperate with the official Napster servers once they
are linked together.  Given that I can't run tcpdump between those servers,
I can't see how they are communicating.  I'm hoping that with the proliferation
of servers outside of napster.com, they might be willing to open up their
specification to make sure all the clients can interoperate.  However, this
server should be compliant with the several currently available Unix Napster
clients.

Main differences from the official napster servers:
* servers can be linked together fully (chat and search)
* channel operators can be defined to moderate specific channels
* source code is freely available

What's included?
* opennap server
* metaserver for directing clients to server groups
* protocol specification
* SOURCE CODE!

There is also a mailing list opennap-admin@lists.sourceforge.net for people
running live servers to share information not directly related to the
development of the server.

* For a description of the napster protocol, please see the file 'napster.txt'
  included with this distribution.

* INSTALLATION

  To install, simply run the `opennap_setup' utility (provided with the binary
  distributions, or built from source).

  The installation process involves picking a directory where to install
  OpenNap and creating an initial account for administering the server.  By
  default, OpenNap will look for its configuration files in:

	C:\opennap-ng			Win32 platforms (Win 9x/NT/2k/XP)

	-or-

	/usr/local/etc/opennap-ng	Unix platforms

Other stateful files (users, bans, channels, etc) are in:

	c:\opennap-ng			Win32 platforms (Win 9x/NT/2k/XP)

	-or-

	/usr/local/var/opennap-ng	Unix platforms

  If you wish to install OpenNap in another location, simply enter the new
  directory when `opennap_setup' prompts you.

  Keep in mind that the above have changed!  All files used to be in
  $prefix/share/opennap-ng and now they are in $prefix/etc/opennap-ng
  and $prefix/var/opennap-ng.  Those updating might need to move files
  or create links depending on your taste.  In docs/scripts/fixfiles
  is a script that will copy the files from their old location to the
  new one as a convenience for you.

  The next step is to create an initial user account for the owner of the
  server.  `opennap_setup' will prompt you for the nickname to use, the
  password for the account and your email address (note that you ARE NOT
  required to enter a real email address if you don't want to--simply
  use the default email@here.com if you like)

  IF you installed OpenNap in a different directory than the default,
  `opennap_setup' will create a config file for you so that opennap knows
  that it is installed in the directory you chose.  However, in order for
  opennap to find this config file, you will have to tell it where the
  config file is when you start the server.  You do this by specifying the
  -c command line option as such:

	opennap -c PATH

  where PATH is the _full_ path to the config file.

  (NOTE: for Win32 users, `opennap_setup' will create a `launch.bat' script
  in the directory you chose to install in which contains the above command
  so that you can simply run the batch file to start the server.  Just make
  sure you copy the opennap.exe executable into that directory first).
  
* Linking Servers

  1.  you need to add the dns name of the server you wish to allow to connect
      to your server in the database.  This needs to be done on both
      ends, as it does mutual authentication to prevent bogus servers from
      being able to join.

	edit /usr/local/etc/opennap-ng/servers (c:\opennap-ng\servers on Win32)
	add:
		<server name> <password> <local_password>

	where <server name> is the dns name of the server, and <password>
	is the password for that server to gain access.  <local_password> is
	the password your server uses to authenticate itself to the other
	server (so the two password fields should be swapped on the
	other server).

   2.  connect to your server and gain elite access, and execute the
       `server connect' command (currently only BWap has support for
       the opennap commands), specifying the host and port to connect to.

       In BWap, you would do:
	/admin connect <server>
       In the windows client or other client which does not natively support
       the opennap extensions, you can do:
	/msg operserv connect <server>

* Server to Server messages

There are many cases where client commands need to be passed to peer servers
in order to maintain database consistency across all servers.  However, most
of the commands that the clients use don't specify the user who performed
the action.  This implementation borrows from the IRC protcol and prefixes
client commands with
	:user
to indicate to peer servers which user issued the command.  Commands which
already specify the user name, such as the login request (2), are not
prefixed.  Each request handler should have a comment at the beginning
specifying what input it expects.

* Extensions to messages

200	search [CLIENT]

	TYPE <mime-type>

		OpenNap supports other media types besides mp3.  By default,
		searches will only match mp3 files.  A client can however
		search for other media types by specifying a
		partial MIME content-type (audio, video, text, application,
		image).  (See message 10300 for adding media to the database).
		The special keyword `any' will match any media type in the
		database.

		The results of the search are returned as with mp3, except
		the fields for bitrate, sample frequency and length are
		meaningless (and are set to 0 in this implmentation).  The
		client can then download as they would any other mp3 file.

	FILENAME EXCLUDES "..."

		allows filtering search results by excluded all files which
		match words in an exclude list.  this must be used in
		conjunction with FILENAME CONTAINS "...".
		(NOTE: as of Napster 2.0 BETA 8, words in the
		FILENAME CONTAINS string which are prefixed with a minus
		sign (-) are considered to be exclusive.  You are better
		off implementing it this way for compatibility.)

	[ SIZE | DURATION ] ["AT LEAST" | "EQUAL TO" | "AT BEST"] <integer>

		allows matching on the file size and song length in addition
		to the other file attributes

612	ban user/ip [CLIENT]

	Format: <user|ip> [ "reason" [timeout] ]

	opennap adds the optional `timeout' fields which specifies the time,
	in seconds, that the ban should be enforced.  Note that if `timeout'
	is present, the `reason' field MUST be present, even if empty.

* Non-standard messages

The following messages are not present in the official napster servers, but
are implemented as additional functionality in the opennap server.

10000	client quit [SERVER]

	<nick>

	the message is sent to peer servers when a client connection has
	closed

10010	server login [SERVER]

	<server-name> <nonce> <compression>

	<server-name>	the dns name of the server wishing to connect
	<nonce>		a random string to use for authentication
	<compression>	the maximum zip (LZ77) compression level

	this message is sent when a connection wishes to identify itself as
	a peer server.  when a server receives this message, it will send its
	own login command to the peer to initial mutual authentication

10011	server login ack [SERVER]

	<hash-response>

	to authenticate itself, the server will hash the value
		<peer-nonce><nonce><server-pass>
	using the MD5 algorithm.  this allows the servers to mutually
	authenticate without the plaintext passwords traversing the network

10012	request server link list [CLIENT]

	Format: <empty>

	client requests the current server link info

10013	user ip [SERVER] (deprecated, this info is now appended to msg 2)

	<user> <ip> <port> <server>

	this message is used for a server to pass the ip address of a
	locally connected client to its peer servers, since that information
	is not available in the login message.  <ip> is an unsigned long
	integer specifying the ip address for <user>

10014	registration info [SERVER]

	<user> <pass> <level> <email> <created> <last-seen>

	<pass>		user's password
	<level>		user's default level
	<email>		user's email address
	<created>	time at which the account was created
	<last-seen>	last time the user logged into a server

	When a server detects that the user table is out of sync between
	servers, it will send its notion of what the entry should look like
	to all the other servers.  If a receiving server has a matching
	user, it checks the <created> time to see which is the oldest
	entry and updates accordingly.

10018	encapsulated message [SERVER]

	:<sender> <recip> <message>

	This command allows a server to send a private message to a user on
	a remote server.  <message> should be a well-formed Napster message,
	complete with tag and length header.

10019	server link info [SERVER]

	<server> <port> <server> <port> <hops>

	This message is used by servers to share information about the
	topology of the linked servers.  when a server joins, a message is
	sent to all other nodes in the cluster.  <hops> is incremented each
	time the message is relayed so that each server knows how far away
	the others are.

10020	server quit [SERVER]

	:<server> <server> "<reason>"

	This message is used to notify other servers in the group that a
	server has delinked.  The server with the peer connection to the
	server that has quit should send this message.

10021	notify mods [SERVER]

	:<server> <loglevel> "<message>"


NEWS
$Id: NEWS 297 2004-08-29 21:14:20Z khaytsus $

Open Souce Napster Server
-------------------------
[opennap-ng 0.48]

Config dirs have changed, you WILL POSSIBLY NEED TO MOVE FILES!  The new
structure is as such:

$prefix/etc/opennap-ng is the new Config Directory, and the following
files are within in:
class, config, filter, block, motd, servers

$prefix/var/opennap-ng is the new State Directory, and the following
files are within it:
bans, channels, opennap-ng.pid, log, level.log, users

docs/scripts/fixfiles is a script to fix this situation for you, or
you may do it yourself.

[opennap-ng 0.47]

Lots of bug fixes, some day I'll get a decent list of bug fixes and new
features.

Added some patches by Hoshi that should improve the servers latency,
speed up searches and logoff hash removals.

[opennap-ng 0.46]

This file highlights the major user visible changes between versions.  For a
detailed description of all changes, please see the ChangeLog file.

[opennap 0.44]

Fixed to speed up the process of removing files from the search index so
that there is no longer a noticeable pause when a user with a large
collection disconnects from the server.

[opennap 0.43]

Added `eject_after' variable to specify how long a client is exempt from
getting killed to make room for other clients after initial login for
not sharing enough files (default: 120 seconds).

Added `index_path_depth' variable to control how many levels of directory
are indexed for searching.  Often the leading parts of the path aren't
useful for searches, so indexing just the last 2 levels provides enough
info without wasting a lot of memory.  (Default: 2)

Added `index_ignore_suffix' variable to control whether filename extensions
are indexed.  (Default: yes)

[opennap 0.42]

stat_server_pass and stat_server_user are now hidden variables, meaning that
they can't be queried via the server.

added -r command line option, and remote_config boolean option to disable
remote configuration of the server.  only the rehash function will cause the
existing configuration to change.  none of the server variables can be
queried or changed when in this mode.

you should no longer need to double quote values in the config file if they
contain spaces (this was a bug in 0.41)

SIGHUP will now cause opennap to re-read its config file

Included registry entry for Windows 98SE to increase the max number of
allowed connectsion up to 1024.  A BIG THANKS to Damian Hodgkiss
<mian@thirty4.com> of the Napigator project for this gem of information.
See the FAQ for more details.

The Win32 port is once again compiled natively using MS Visual C++ instead of
the CYGWIN toolkit.

Fixed crash in search_result_end which could be triggered by a user.
Missing a call to check for a server issuing the command.

[opennap 0.41]

Improved Napigator support.  Added new variables report_name, report_ip, and
report_port which set the information reported to Napigator.  These default
from the other configuration variables and only need to be set if OpenNap
doesn't guess the correct values.

[opennap 0.40]

Added a `block' configuration file which contains a list of regular
expressions.  Any file a client attempts to share that matches one of the
regexps will be discarded and unavailable for download via the server.  See
sample.block for a useful list to begin with.

Added `allow_share' boolean configuration options which controls whether or
not clients are allowed to share files.  Requests to share files are
silently ignored.

Added `min_read' configuration variable to allow a low-watermark log
message when less than this many bytes are read from a server link.  Useful
for detecting slow servers.

Added support for directly reporting server stats to a Napigator-style
server.

[opennap 0.39]

The metaserver can now read the servers list from a file as well as from the
command line.

Added `server_chunk' config variable to control the minimum packet size to
be sent between servers (Default: 0 = no minimum).

Added `ghost_kill' boolean to allow disabling the ghost kill feature.

Added `auto_link' boolean to make the server automatically attempt to link
to all servers upon startup.

Added support for ircd-style i:lines (allowed ips), e:lines (exempt ips),
d:lines (drop without any notification)

Added `login_interval' and `register_interval' variables to control how fast
a client from the same ip may log in or register a nickname.

Support for Napster BETA 8 and BETA 9.

Added `remote_browse' boolean variable to disable remote browsing.

Added support for histogram of number and bytecount for each message type
received.

Disabled the 10301 (`new' browse method) [Note: a few clients depended on
this, but should have been updated to use BETA8+ method of direct client to
client browsing.]

Added `max_searches' variable to limit the number of pending searches a
client can have.

Added usermode -msg to disable private messages.

Added usermode -whois to disable whois notifications for mod+ users.

[opennap 0.38]

***NOTE*** Because of a change in the server quit messages, if 0.38 is used

src/ban.c
/* Copyright (C) 2000-1 drscholl@users.sourceforge.net
This is free software distributed under the terms of the
GNU Public License.  See the file COPYING for details.

$Id: ban.c 436 2006-09-04 14:56:32Z reech $ */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <sys/types.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <ctype.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <limits.h>
#ifndef WIN32
# include <unistd.h>
#endif
#include "opennap.h"
#include "debug.h"

void ban_user_internal(CONNECTION *con, char *user, time_t btimeout, char *reason)
{
    BAN    *b;
    LIST   *list;
    char   *banptr, realban[256];
    time_t  timeout = 0;

    (void) con;

    ASSERT(validate_connection(con));
    timeout = btimeout;

    truncate_reason(reason);

    /* Normalize to "user!*" */
    banptr = normalize_ban(user, realban, sizeof(realban));

    /* check to see if this user is already banned */
    for (list = global.banList; list; list = list->next)
    {
        b = list->data;
        if(!strcasecmp(banptr, b->target))
        {
            log_message_level(LOG_LEVEL_ERROR, "ban_user_internal: %s (%s) already banned", realban, banptr);
            return;
        }
    }
    b=create_ban(banptr, global.serverName, reason, global.current_time, btimeout);
    if(!b) 
    {
        log_message_level(LOG_LEVEL_ERROR,"ban_user_internal: Out of memory when banning %s", banptr);
        return;
    }

    notify_mods(BANLOG_MODE, "%s%s banned %s%s%lu%s: %s", "Server ", global.serverName, b->target, (timeout > 0) ? " for " : "", (timeout > 0) ? timeout : 0, (timeout > 0) ? " seconds" : "", b->reason);
    pass_message_args(con, MSG_CLIENT_BAN, ":%s %s \"%s\" %d",global.serverName, user, reason, timeout); 

}

/* Create a ban record and initializes all vars. Returns Pointer to 
record if successful - NULL if not. */
BAN *create_ban(char *target, char *issuer, char *reason, time_t starttime, time_t timeout) 
{
    BAN *b;
    LIST *list;

    list=NULL;

    while (1)
    {
        /* create structure and add to global ban list */
        b = CALLOC(1, sizeof(BAN));
        if(!b)
            break;
        strncpy(b->target, target, sizeof(b->target) - 1);
        b->target[sizeof(b->target) - 1] = 0;
        strncpy(b->setby, issuer, sizeof(b->setby) - 1);
        b->setby[sizeof(b->setby) - 1] = 0;
        strncpy(b->reason, reason, sizeof(b->reason) - 1);
        b->reason[sizeof(b->reason) - 1] = 0;
        b->when = starttime;
        b->timeout = timeout;
        b->when_deleted = 0;
        list = CALLOC(1, sizeof(LIST));
        if(!list) 
		{
            OUTOFMEMORY("ban");
            break;
        }
        list->data = b;
        list->next = global.banList;
        global.banList = list;
        return b;
    };

    /* we only get here on error */
    OUTOFMEMORY("ban");
    free_ban(b);
    if(list)
        FREE(list);
    return NULL;
}


void free_ban(BAN * b)
{
    if(b)
    {
        //if(b->target)
        //    FREE(b->target);
        //if(b->setby)
        //    FREE(b->setby);
        //if(b->reason)
        //    FREE(b->reason);
        FREE(b);
    }
}

char *normalize_ban( char *src, char *dest, int destlen)
{
    int is_ip = 1;
    char *s;
    /* normalize the ban to the full user!host syntax */
    if(strchr(src, '!')) 
    {
        snprintf(dest, destlen, "%s", src);
    } 
    else 
    {
        s = src;
        while (*s)
        {
            if(!(isdigit(*s) || *s == '.'))
                is_ip = 0;
            s++;
        }
        if(is_ip) 
        {
            char   *star;

            /* append a star if the last char is a . so that it means the same
            * as the old-style ban
            */
            if(*src && src[strlen(src) - 1] == '.')
                star = "*";
            else
                star = "";
            snprintf(dest, destlen, "*!%s%s", src, star);  /* must be an ip/dns name? */
        }
        else 
        {
            snprintf(dest, destlen, "%s!*", src);  /* must be a nick */
        }
    }
    /* log_message_level(LOG_LEVEL_DEBUG,"normalize_ban: (%s) --> (%s)", src, dest); */
    return dest;

}

/* 612 [ :<sender> ] <user!ip> [ "<reason>" [ time [starttime] ] ] */
HANDLER(ban)
{
    BAN    *b;
    LIST   *list;
    int     ac = -1;
    char   *av[4], *sendernick;
    char   *banptr, realban[256];
    USER   *sender;
    int     timeout = 0;
    time_t  starttime = 0;

    (void) len;
    ASSERT(validate_connection(con));

    if(pop_user_server(con, tag, &pkt, &sendernick, &sender))
        return;

    if(sender && sender->level < LEVEL_MODERATOR)
    {
        permission_denied(con);
        return;
    }

    if(pkt) 
    {
        /* Do some logging on the server tban sync thing ...  */
        ac = split_line(av, sizeof(av) / sizeof(char *), pkt);
    }
    if(ac < 1)
    {
        unparsable(con);
        return;
    }

    /* When we have 3 args then it's a tban with or without a start time.
    First the timeout to stay compatible to the old semantics */
    if(ac >= 3)
    {
        timeout = atoi(av[2]);
        if(timeout < 0)
        {
            if(ISUSER(con))
                send_cmd(con, MSG_SERVER_NOSUCH, "invalid ban timeout");
            return;
        }
    }

    /* And only if we have 4 args then the new semantics is used.
    The 4th arg is the ban starttime. */
    if( ac == 4 ) 
    {
        starttime=atol(av[3]);
        if(starttime < 0) 
        {
            if(ISUSER(con))
                send_cmd(con, MSG_SERVER_NOSUCH, "invalid ban starttime");
            return;
        }
    } 
    else 
    {
        starttime=global.current_time;
    }

    /* Do some logging on the server tban sync thing ... */
    if( ! ISUSER(con)) 
    {
        log_message_level(LOG_LEVEL_DEBUG, "ban: Server %s issued %d args: \"%s\" \"%s\" \"%s\" \"%s\"", sendernick, ac, av[0], ac>1?av[1]:"", ac>2?av[2]:"",ac>3?av[3]:"");
    }

    banptr = normalize_ban(av[0], realban, sizeof(realban));

    /* check to see if this user is already banned */
    for (list = global.banList; list; list = list->next)
    {
        b = list->data;
        /* If the user is already banned then look for the start time 
        of the ban. If our ban is older ( means smaller ) then ignore the ban 
        request. If our ban is younger then we have to update 
        the start time of the ban to reflect that this user
        had been previously banned on another server.
        This should avoid stale tbans without updating a 
        deletion stub in the banlist and all the hassles around it.
        As unbans are routed meanwhile these stale tbans should disappear quite 
        quickly. */
        if(!strcasecmp(banptr, b->target))
        {
            if(ISUSER(con)) 
            {
                send_cmd(con, MSG_SERVER_NOSUCH, "already banned");
            } 
            else 
            {
                if( b->when > starttime ) 
                {
                    b->when=starttime;
                }
            }
            return;
        }
    }

    if(ac > 1)
        truncate_reason(av[1]);

    b=create_ban(banptr, sendernick, ac > 1 ? av[1] : "", starttime, timeout);
    if( !b ) 
    {
        log_message_level(LOG_LEVEL_ERROR,"ban: Out of memory when banning %s", banptr);
        return;
    }

    notify_mods(BANLOG_MODE, "%s%s banned %s%s%s%s: %s",
        !sender ? "Server " : "", sendernick, b->target, 
        (timeout > 0) ? " for " : "",
        (timeout > 0) ? av[2] : "",
        (timeout > 0) ? " seconds" : "", b->reason);
    pass_message_args(con, tag, ":%s %s \"%s\" %d %lu", sendernick, av[0], ac > 1 ? av[1] : "", timeout, starttime);

}

/* 614 [ :<sender> ] <nick!ip> [ "<reason>" ] */
HANDLER(unban)
{
    /*    USER   *user; */
    USER   *sender;
    LIST  **list, *tmpList;
    BAN    *b;
    int     ac = -1;
    char   *av[2], *sendernick;
    char   *banptr, realban[256];

    /*    (void) tag; */
    (void) len;

    ASSERT(validate_connection(con));

    /* Previously the sender had been just a user causing 
    messages like: pop_user: (tag 614) could not find user
    This means that servers had not been able to unban users. */
    if(pop_user_server(con, tag, &pkt, &sendernick, &sender))
        return;

    /*    if(pop_user(con, &pkt, &user) != 0)
    return;
    */

    if(pkt)
        ac = split_line(av, FIELDS(av), pkt);
    if(ac < 1)
    {
        unparsable(con);
        return;
    }

    /* If the sender is null then the unban has been a server unban. */
    if(sender && sender->level < LEVEL_MODERATOR)
    {
        permission_denied(con);
        return;
    }


    banptr = normalize_ban(av[0], realban, sizeof(realban));
    if(ac > 1)
        truncate_reason(av[1]);

    /* Have some overview over freshly generated code ... 
    Parameters listed: Unbanning server, Unbanned user, and reason. */
    if( ! sender ) 
    {
        log_message_level(LOG_LEVEL_DEBUG,"unban: Server %s unbanned %s: %s", sendernick, banptr, ac > 1 ? av[1] : "");
    }

    for (list = &global.banList; *list; list = &(*list)->next)
    {
        b = (*list)->data;
        if(!strcasecmp(banptr, b->target))
        {
            tmpList = *list;
            *list = (*list)->next;
            FREE(tmpList);
            /* To prevent coredumps when the unbanning user is a server ... */
            if( sender ) 
            {
                notify_mods(BANLOG_MODE, "%s removed ban on %s: %s",sender->nick, b->target, ac > 1 ? av[1] : "");
                pass_message_args(con, tag, ":%s %s \"%s\"", sender->nick,b->target, ac > 1 ? av[1] : "");
            } 
            else 
            {
                notify_mods(BANLOG_MODE, "%s removed ban on %s: %s",sendernick, b->target, ac > 1 ? av[1] : "");
                pass_message_args(con, tag, ":%s %s \"%s\"", sendernick, b->target, ac > 1 ? av[1] : "");
            }
            free_ban(b);
            return;
        }
    }
    if(ISUSER(con))
        send_cmd(con, MSG_SERVER_NOSUCH, "no such ban");
}

/* 615 */
/* show the list of current bans on the server */
HANDLER(banlist)
{
    LIST   *list;
    BAN    *ban;

    (void) tag;
    (void) len;
    (void) pkt;
    if(ISUSER(con))
    {
        /* The banlist is mod+ only from this point on. */
        if(con->user->level < LEVEL_MODERATOR)
        {
            permission_denied(con);
            return;
        }
    }
    ASSERT(validate_connection(con));
    CHECK_USER_CLASS("banlist");
    for (list = global.banList; list; list = list->next)
    {
        ban = list->data;
        send_cmd(con, MSG_SERVER_IP_BANLIST, "%s %s \"%s\" %u %d", ban->target, ban->setby, ban->reason, ban->when, ban->timeout);
    }
    /* terminate the banlist */
    send_cmd(con, MSG_CLIENT_BANLIST, "");
}

int check_ban(CONNECTION * con, const char *nick, const char *host)
{
    LIST   *list;
    BAN    *ban;
    char    mask[256];
    int     deltat;
    int     howmany;

    snprintf(mask, sizeof(mask), "%s!%s", nick, host);
    for (list = global.banList; list; list = list->next)
    {
        ban = list->data;
        if((ban->timeout == 0 || ban->when + ban->timeout > global.current_time) && glob_match(ban->target, mask))
        {
            notify_mods(BANLOG_MODE,"Connection from %s: %s banned: %s",mask, ban->target, NONULL(ban->reason));
            if( ! ban->connect_counter ) 
            {
                ban->firstconnect=global.current_time;
            }
            ban->connect_counter++;

            /* Calculate the time since login in minutes */
            deltat=( global.current_time - ban->firstconnect ) / 60;
            howmany=deltat?ban->connect_counter / deltat:0;

            log_message_level(LOG_LEVEL_LOGIN, "check_ban: %s ( %s ) on %s connected %d times ( %d per minute )",nick,ban->target,host,ban->connect_counter,howmany);

            if(ISUNKNOWN(con)) 
            {
				stats.login_ce_banned++;
                ibl_kill(con, MSG_SERVER_ERROR, "%s banned: %s", ban->target, NONULL(ban->reason));
            }
            else if(ISSERVER(con)) 
            {
                /* issue a kill to remove this banned user */
                pass_message_args(con, MSG_CLIENT_KILL, ":%s %s %s banned: %s", global.serverName, nick, ban->target, NONULL(ban->reason));
                notify_mods(KILLLOG_MODE, "Server %s killed %s: %s banned: %s", global.serverName, nick, ban->target, NONULL(ban->reason));
            }
            return 1;
        }
    }
    return 0;
}

int save_bans(void)
{
    int     fd;
    LIST   *list;
    BAN    *b;
    char    path[_POSIX_PATH_MAX], tmppath[_POSIX_PATH_MAX];
    char    inbuf[1024];
    struct  stat sts;

    log_message_level(LOG_LEVEL_SERVER, "save_bans: dumping ban database" );

    if(global.banList) 
    {
        snprintf(tmppath, sizeof(tmppath), "%s/bans.tmp", global.varDir);
        if((fd = open(tmppath, O_WRONLY | O_CREAT , S_IRUSR | S_IWUSR)) == -1) 
        {
            logerr("save_bans", tmppath);
            return -1;
        }
        for (list = global.banList; list; list = list->next) 
        {
            b = list->data;
            /* As the b->reason is determined by max_reason ( a configurable var ) we must 
            not determine the strings via format specifiers. The same goes for targets, issuers.
            */
            snprintf(inbuf,sizeof(inbuf)-2,"%s %s %lu \"%s\" %d", b->target, b->setby,b->when, b->reason, b->timeout);
            strcat( inbuf, LE );

            fake_fputs( inbuf, fd );
        }
        if(close(fd)) 
        {
            logerr("save_bans", "close");
            return -1;
        }
        snprintf(path, sizeof(path), "%s/bans", global.varDir);
        if(stat(path, &sts) == -1 && errno == ENOENT)
        {
            log_message_level(LOG_LEVEL_DEBUG, "%s file does not exist\n", path);
        }
        else 
        {
            if(unlink(path))
                logerr("save_bans", "unlink");       /* not fatal, may not exist */
        }
        if(rename(tmppath, path)) 
        {
            logerr("save_bans", "rename");
            return -1;
        }
    }                                    
    return 0;
}

int load_bans(void)
{
    int    fd;
    LIST   *list, **last = &global.banList, *p;
    BAN    *b;
    int     ac;
    char   *av[5], path[_POSIX_PATH_MAX];
    char   *banptr, realban[256];
    int    dupban;

    snprintf(path, sizeof(path), "%s/bans", global.varDir);
    if((fd = open(path, O_RDONLY))==-1)
    {
        if(errno != ENOENT)
            logerr("load_bans", path);
        return -1;
    }
    while (fake_fgets(Buf, sizeof(Buf)-1, fd))
    {
        ac = split_line(av, FIELDS(av), Buf);
        if(ac < 1)
            continue;
        banptr = normalize_ban(av[0], realban, sizeof(realban));

        /* check to see if this user is already banned */
        dupban=0;
        for (p = global.banList; p; p = p->next) 
        {
            b = p->data;
            dupban = dupban || (!strcasecmp(banptr, b->target));
        }
        if( dupban )  /* This ban entry had been read before - don't create it again. */
            continue;

        b = CALLOC(1, sizeof(BAN));
        if(!b)
        {
            OUTOFMEMORY("load_bans");
            close(fd);
            return -1;
        }
        strncpy(b->target, banptr, sizeof(b->target) - 1);
        b->target[sizeof(b->target) - 1] = 0;
        if(ac >= 4)
        {
            strncpy(b->setby,av[1], sizeof(b->setby) - 1);
            b->setby[sizeof(b->setby) - 1] = 0;
            b->when = atol(av[2]);
            truncate_reason(av[3]);
            strncpy(b->reason, av[3], sizeof(b->reason) - 1);
            b->reason[sizeof(b->reason) - 1] = 0;
            if(ac > 4)
                b->timeout = atoi(av[4]);
        }
        else
        {
            /* old user ban style */
            strncpy(b->setby, global.serverName, sizeof(b->setby) - 1);
            b->setby[sizeof(b->setby) - 1] = 0;
            b->reason[0] =  0;
            b->when = global.current_time;
        }
        list = CALLOC(1, sizeof(LIST));
        if(!list)
        {
            OUTOFMEMORY("load_bans");
            free_ban(b);
            close(fd);
            return -1;
        }
        b->connect_counter=0;
        list->data = b;
        /* keep the bans in the same order (roughly reverse chronological) */
        *last = list;
        last = &list->next;
    }
    close(fd);
    return 0;
}

/* reap expired bans from the list */
void expire_bans(void)
{
    LIST  **list, *tmp;
    BAN    *b;

    list = &global.banList;
    while (*list)
    {
        b = (*list)->data;
        if(b->timeout > 0 && b->when + b->timeout < global.current_time)
        {
            tmp = *list;
            *list = (*list)->next;
            FREE(tmp);
            /* make sure all servers are synched up */
            pass_message_args(NULL, MSG_CLIENT_UNBAN, ":%s %s \"expired after %d seconds\"", global.serverName, b->target, b->timeout);
            notify_mods(BANLOG_MODE, "%s removed ban on %s: expired after %d seconds", global.serverName, b->target, b->timeout);
            free_ban(b);
            continue;
        }
        list = &(*list)->next;
    }
}


int ibl_check(unsigned int ip) 
{
    LIST   *list;
    BAN    *b;
    char   *cip;

    if(global.ibl_ttl) 
    {
        cip = my_ntoa(BSWAP32(ip));
        for (list = global.internalBanList; list; list = list->next) 
        {
            b = list->data;
            if(!strcasecmp(cip, b->target)) 
            {
#ifdef ONAP_DEBUG
                log_message_level( LOG_LEVEL_DEBUG, "ibl_check: %s is banned: %s", b->target, b->reason);
#endif
                return 1;
            }
        }
    }
    return 0;
}

int ibl_kill(CONNECTION * con, unsigned int msgtype, const char *fmt, ...) 
{
    LIST   *list;
    BAN    *b;
    char   *realban, reason[256] ;
    va_list ap;

    (void) con;
    ASSERT(validate_connection(con));

    va_start(ap, fmt);
    vsnprintf(reason, sizeof(reason), fmt, ap);
    va_end(ap);
    realban = my_ntoa(BSWAP32(con->ip));

    send_cmd( con, msgtype, "%s", reason);
    destroy_connection(con);

    if(global.ibl_ttl) 
    {
        log_message_level( LOG_LEVEL_LOGIN, "ibl_kill: %s %s", realban, reason);
        /* check if already banned */ 
        for (list = global.internalBanList; list; list = list->next) 
        {
            b = list->data;
            if(!strcasecmp(realban, b->target)) 
            {
#ifdef ONAP_DEBUG
                log_message_level(LOG_LEVEL_DEBUG, "ibl_kill: %s already banned", realban);
#endif
                return 1;
            }
        }
        while (1)
        {
            b = CALLOC(1, sizeof(BAN));
            if(!b)
                break;
            strncpy(b->target,realban, sizeof(b->target) - 1);
            b->target[sizeof(b->target) - 1] = 0;
            strncpy(b->reason, reason, sizeof(b->reason) - 1);
            b->reason[ sizeof(b->reason) - 1] = 0;
            b->when = global.current_time;
            b->timeout = global.ibl_ttl;
            list = CALLOC(1, sizeof(LIST));
            if(!list)
                break;
            list->data = b;
            list->next = global.internalBanList;
            global.internalBanList = list;

            return 0;
        };
        /* we only get here on error */
        OUTOFMEMORY("ibl_kill");
        if(b)
            free_ban(b);
        if(list)
            FREE(list);
        return -1;
    } 
    else 
    {
        log_message_level( LOG_LEVEL_LOGIN, "login: killed %s %s", realban, reason);
    }
    return 0;
}

void ibl_expire(void) 
{
    LIST  **list, *tmp;
    BAN    *b;
    int     expired = 0;

    if(stats.ibl_db || global.ibl_ttl) 
    {
        stats.ibl_db = 0;
        list = &global.internalBanList;
        while (*list) 
        {
            b = (*list)->data;
            if(b->timeout > 0 && b->when + b->timeout < global.current_time) 
            {
                log_message_level(LOG_LEVEL_DEBUG, "ibl_expire: %s", b->target);
                expired++;
                tmp = *list;
                *list = (*list)->next;
                FREE(tmp);
                free_ban(b);
                continue;
            } 
            else 
            {
                stats.ibl_db++;
            }
            list = &(*list)->next;
        }
        log_message_level(LOG_LEVEL_SERVER, "ibl_expire: %u expired, %u in internal ban list", expired, stats.ibl_db);
    }
}
